controlplane/kubeadm/internal/etcd_client_generator.go

```diff
@@ -72,26 +78,48 @@ func (c *EtcdClientGenerator) forFirstAvailableNode(ctx context.Context, nodeNam
 
 // forLeader takes a list of nodes and returns a client to the leader node.
 func (c *EtcdClientGenerator) forLeader(ctx context.Context, nodeNames []string) (*etcd.Client, error) {
-	var errs []error
+	// This is an additional safeguard for avoiding this func to return nil, nil.
+	if len(nodeNames) == 0 {
+		return nil, errors.New("invalid argument: forLeader can't be called with an empty list of nodes")
+	}
 
+	// Loop through the existing control plane nodes.
+	var errs = make([]error, 0)
 	for _, nodeName := range nodeNames {
+		// Get a temporary client to the etcd instance hosted on the node.
 		client, err := c.forFirstAvailableNode(ctx, []string{nodeName})
 		if err != nil {
 			errs = append(errs, err)
 			continue
 		}
 		defer client.Close()
+
+		// Get the list of members.
 		members, err := client.Members(ctx)
 		if err != nil {
 			errs = append(errs, err)
 			continue
 		}
+
+		// Loop through the list of etcd members reported checking for the current leader.
 		for _, member := range members {
```

Comment:
The old code only returned a client for the leader if it identified itself as leader. The new code potentially returns the leader based on the information of a follower. Is that intended?

Reply:
yes.
this will save some access to etcd